src/txn/txn.c

```diff
@@ -375,10 +375,260 @@ __wt_txn_update_oldest(WT_SESSION_IMPL *session, uint32_t flags)
 #endif
 	}
 
-done:	__wt_writeunlock(session, &txn_global->scan_rwlock);
+done:	__wt_writeunlock(session, &txn_global->oldest_rwlock);
 	return (ret);
 }
 
+#if TIMESTAMP_SIZE > 0
+/*
+ * __txn_set_timestamp --
+ *	Decodes and sets a timestamp.
+ */
+static int
+__txn_set_timestamp(WT_SESSION_IMPL *session,
+     const char *name, uint8_t *timestamp, WT_CONFIG_ITEM *cval)
+{
+	WT_DECL_RET;
+	WT_ITEM ts;
+	char padbuf[2 * TIMESTAMP_SIZE], tsbuf[TIMESTAMP_SIZE];
+	const char *hexts;
+	size_t hexlen;
+
+	if (cval->len == 0) {
+		memset(timestamp, 0, TIMESTAMP_SIZE);
+		return (0);
+	}
+
+	/* Protect against unexpectedly long hex strings. */
+	if (cval->len > 2 * TIMESTAMP_SIZE)
+		WT_RET_MSG(session, EINVAL,
+		    "Failed to parse %s timestamp '%.*s': too long",
+		    name, (int)cval->len, cval->str);
+
+	/*
+	 * The decoding function assumes it is decoding data produced by dump
+	 * and so requires an even number of hex digits.
+	 */
+	if ((cval->len & 1) == 0) {
+		hexts = cval->str;
+		hexlen = cval->len;
+	} else {
+		padbuf[0] = '0';
+		memcpy(padbuf + 1, cval->str, cval->len);
+		hexts = padbuf;
+		hexlen = cval->len + 1;
+	}
+
+	/* Avoid memory allocation to decode timestamps. */
+	ts.data = ts.mem = tsbuf;
+	ts.memsize = sizeof(tsbuf);
+
+	if ((ret = __wt_nhex_to_raw(session, hexts, hexlen, &ts)) != 0)
+		WT_RET_MSG(session, ret, "Failed to parse %s timestamp '%.*s'",
+		    name, (int)cval->len, cval->str);
+	WT_ASSERT(session, ts.size <= TIMESTAMP_SIZE);
+
+	if (ts.size < TIMESTAMP_SIZE)
+		memset(timestamp, 0, TIMESTAMP_SIZE - ts.size);
+
+	/* Copy the raw value to the end of the timestamp. */
+	memcpy(timestamp + TIMESTAMP_SIZE - ts.size,
+	    ts.data, ts.size);
+
+	if (__wt_ts_iszero(timestamp))
+		WT_RET_MSG(session, EINVAL,
+		    "Failed to parse %s timestamp '%.*s': zero not permitted",
+		    name, (int)cval->len, cval->str);
+
+	return (ret);
+}
+
+/*
+ * __txn_global_query_timestamp --
+ *	Query a timestamp.
+ */
+static int
+__txn_global_query_timestamp(
+    WT_SESSION_IMPL *session, uint8_t *ts, const char *cfg[])
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_CONFIG_ITEM cval;
+	WT_TXN_GLOBAL *txn_global;
+	WT_TXN_STATE *s;
+	uint32_t i, session_cnt;
+
+	conn = S2C(session);
+	txn_global = &S2C(session)->txn_global;
+
+	WT_RET(__wt_config_gets(session, cfg, "get", &cval));
+	if (WT_STRING_MATCH("all_committed", cval.str, cval.len)) {
```

Comment:
I think it'd be nice if these functions were pulled into a different `txn_timestamp.c` source file, I also think it'd be nice in this particular function to split the config parsing from the logic.